concordium-cis2/src/lib.rs

```diff
@@ -566,6 +572,162 @@ token_amount_wrapper!(TokenAmountU32, u32);
 token_amount_wrapper!(TokenAmountU16, u16);
 token_amount_wrapper!(TokenAmountU8, u8);
 
+#[cfg(feature = "u256_amount")]
+mod u256_token {
+    use super::*;
+    use primitive_types::U256;
+    #[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Default)]
+    #[repr(transparent)]
+    pub struct TokenAmountU256(pub U256);
+
+    impl ops::Add<Self> for TokenAmountU256 {
+        type Output = Self;
+
+        fn add(self, rhs: Self) -> Self::Output { TokenAmountU256(self.0 + rhs.0) }
+    }
+
+    impl ops::AddAssign for TokenAmountU256 {
+        fn add_assign(&mut self, other: Self) { *self = *self + other; }
+    }
+
+    impl ops::Sub<Self> for TokenAmountU256 {
+        type Output = Self;
+
+        fn sub(self, rhs: Self) -> Self::Output { TokenAmountU256(self.0 - rhs.0) }
+    }
+
+    impl ops::SubAssign for TokenAmountU256 {
+        fn sub_assign(&mut self, other: Self) { *self = *self - other; }
+    }
+
+    impl ops::Mul<U256> for TokenAmountU256 {
+        type Output = Self;
+
+        fn mul(self, rhs: U256) -> Self::Output { TokenAmountU256(self.0 * rhs) }
+    }
+
+    impl ops::Mul<TokenAmountU256> for U256 {
+        type Output = TokenAmountU256;
+
+        fn mul(self, rhs: TokenAmountU256) -> Self::Output { TokenAmountU256(self * rhs.0) }
+    }
+
+    impl ops::MulAssign<U256> for TokenAmountU256 {
+        fn mul_assign(&mut self, other: U256) { *self = *self * other; }
+    }
+
+    impl ops::Rem<U256> for TokenAmountU256 {
+        type Output = Self;
+
+        fn rem(self, other: U256) -> Self::Output { TokenAmountU256(self.0 % other) }
+    }
+
+    impl ops::RemAssign<U256> for TokenAmountU256 {
+        fn rem_assign(&mut self, other: U256) { *self = *self % other; }
+    }
+
+    impl iter::Sum for TokenAmountU256 {
+        fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+            iter.fold(TokenAmountU256(U256::zero()), ops::Add::add)
+        }
+    }
+
+    impl IsTokenAmount for TokenAmountU256 {}
+
+    /// Uses the ULeb128 encoding with up to 37 bytes for the encoding as
+    /// according to CIS-2 specification.
+    impl schema::SchemaType for TokenAmountU256 {
+        fn get_type() -> schema::Type { schema::Type::ULeb128(37) }
+    }
+
+    impl Serial for TokenAmountU256 {
+        fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+            let mut value = self.0;
+            loop {
+                let mut byte = (value.low_u32() as u8) & 0b0111_1111;
+                value >>= 7;
+                if value != <U256>::from(0u8) {
```

Comment:
Maybe use the constant function for zero instead
```suggestion
                if value != U256::zero() {
```